From 8ab274dca5ed87041dff3293c7cb67bb628aeed2 Mon Sep 17 00:00:00 2001
From: Pavel Tumakaev <p.tumakaev@lgepartner.com>
Date: Wed, 22 May 2019 17:32:25 +0300
Subject: [PATCH] Remove null pointer checks for "this" from
 QQmlContextData::resolvedUrl()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

According to ISO/IEC 14882 §5.2.5/3 "If E1 has the type “pointer to class
X,” then the expression E1->E2 is converted to the equivalent form
(*(E1)).E2;". Thus, as QQmlContextData::resolvedUrl() is non-static method,
it cannot be called on a null pointer because it leads to undefined
behavior, and “this“ in QQmlContextData::resolvedUrl() cannot be a null
pointer. According to this document:
https://gcc.gnu.org/gcc-6/porting_to.html , starting from version 6, GCC
optimizations remove null pointer checks for "this", since "the this
pointer can never be null, which is guaranteed by the language rules."
Thus, on the one hand the “if (ctxt)“ check in
QQmlContextData::resolvedUrl() does nothing, on the other “if (engine“
check leads to undefined behavior if ctxt/this == nullptr.

Task-number: QTBUG-75983
Change-Id: Idfb1e26758d83223bb0845139d63e2e8e80dc714

Upstream-Status: Backport [https://codereview.qt-project.org/c/qt/qtdeclarative/+/262573 (dev/5.13)]
---
 src/qml/qml/qqmlcontext.cpp | 27 ++++++++++++---------------
 1 file changed, 12 insertions(+), 15 deletions(-)

diff --git a/src/qml/qml/qqmlcontext.cpp b/src/qml/qml/qqmlcontext.cpp
index 3710cee16..8688346b5 100644
--- a/src/qml/qml/qqmlcontext.cpp
+++ b/src/qml/qml/qqmlcontext.cpp
@@ -438,23 +438,20 @@ QUrl QQmlContext::resolvedUrl(const QUrl &src)
 
 QUrl QQmlContextData::resolvedUrl(const QUrl &src)
 {
-    QQmlContextData *ctxt = this;
-
     QUrl resolved;
     if (src.isRelative() && !src.isEmpty()) {
-        if (ctxt) {
-            while(ctxt) {
-                if (ctxt->url().isValid())
-                    break;
-                else
-                    ctxt = ctxt->parent;
-            }
-
-            if (ctxt)
-                resolved = ctxt->url().resolved(src);
-            else if (engine)
-                resolved = engine->baseUrl().resolved(src);
-        }
+        QQmlContextData *ctxt = this;
+        do {
+            if (ctxt->url().isValid())
+                break;
+            else
+                ctxt = ctxt->parent;
+        } while (ctxt);
+
+        if (ctxt)
+            resolved = ctxt->url().resolved(src);
+        else if (engine)
+            resolved = engine->baseUrl().resolved(src);
     } else {
         resolved = src;
     }
